python/ray/tests/accelerators/test_hpu.py

```diff
@@ -113,6 +114,256 @@ def test_check_accelerator_info():
+                tuple(self.hpu_ids),
+            )
+
+    # Create one actor per HPU.
+    actors = [Actor1.remote() for _ in range(num_nodes * num_hpus_per_raylet)]
+    # Make sure that no two actors are assigned to the same HPU.
+    locations_and_ids = ray.get(
+        [actor.get_location_and_ids.remote() for actor in actors]
+    )
+    node_names = {location for location, hpu_id in locations_and_ids}
+    assert len(node_names) == num_nodes
+    location_actor_combinations = []
+    for node_name in node_names:
+        for hpu_id in range(num_hpus_per_raylet):
+            location_actor_combinations.append((node_name, (f'{hpu_id}',)))
+
+    assert set(locations_and_ids) == set(location_actor_combinations)
+
+    # Creating a new actor should fail because all of the HPUs are being
+    # used.
+    a = Actor1.remote()
+    ready_ids, _ = ray.wait([a.get_location_and_ids.remote()], timeout=0.01)
+    assert ready_ids == []
+
+
+def test_actor_habana_visible_devices(shutdown_only):
+    """Test user can overwrite HABANA_VISIBLE_MODULES
+    after the actor is created."""
+    ray.init(resources={'HPU': 1})
+
+    @ray.remote(resources={'HPU': 1})
+    class Actor:
+        def set_habana_visible_devices(self, habana_visible_devices):
+            os.environ["HABANA_VISIBLE_MODULES"] = habana_visible_devices
+
+        def get_habana_visible_devices(self):
+            return os.environ["HABANA_VISIBLE_MODULES"]
+
+    actor = Actor.remote()
+    assert ray.get(actor.get_habana_visible_devices.remote()) == "0"
+    ray.get(actor.set_habana_visible_devices.remote("0,1"))
+    assert ray.get(actor.get_habana_visible_devices.remote()) == "0,1"
+
+
+def test_hpu_ids(shutdown_only):
+    num_hpus = 3
+    ray.init(num_cpus=num_hpus, resources={"HPU": num_hpus})
+
+    def get_hpu_ids(hpus_per_worker):
+        hpu_ids = ray.get_runtime_context().get_accelerator_ids()["HPU"]
+        assert len(hpu_ids) == hpus_per_worker
+        modules = os.environ.get("HABANA_VISIBLE_MODULES")
+        if modules is not None:
+            assert modules == ",".join([str(i) for i in hpu_ids])  # noqa
+        for hpu_id in hpu_ids:
+            assert hpu_id in [str(i) for i in range(num_hpus)]
+        return hpu_ids
+
+    f0 = ray.remote(resources={"HPU": 0})(lambda: get_hpu_ids(0))
+    f1 = ray.remote(resources={"HPU": 1})(lambda: get_hpu_ids(1))
+    f2 = ray.remote(resources={"HPU": 2})(lambda: get_hpu_ids(2))
+
+    # Wait for all workers to start up.
+    @ray.remote
+    def g():
+        time.sleep(0.2)
+        return os.getpid()
+
+    start_time = time.time()
+    while True:
+        num_workers_started = len(set(ray.get([g.remote() for _ in range(num_hpus)])))
+        if num_workers_started == num_hpus:
+            break
+        if time.time() > start_time + 10:
+            raise RayTestTimeoutException(
+                "Timed out while waiting for workers to start up."
+            )
```

Comment:
Not sure what this code is doing and why we need it

Reply:
This test is modified from tests for GPU and neurial core in python/ray/tests/test_advanced_2.py